src/Workspaces/Core/Portable/Shared/TestHooks/AsynchronousOperationListener.cs

```diff
@@ -5,14 +5,16 @@
 using System.Collections.Immutable;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
+using System.Threading;
 using System.Threading.Tasks;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Shared.TestHooks
 {
     internal sealed partial class AsynchronousOperationListener : IAsynchronousOperationListener, IAsynchronousOperationWaiter
     {
-        private readonly object _gate = new object();
+        private readonly NonReentrantLock _gate = new NonReentrantLock();
```

Comment:
I've still seen stack where this will run code while waiting, so if you think you're changing this to help that...I don't think it will.

Reply:
Never mind, @heejaechang did want this to prevent acquiring the lock twice.